### PR TITLE
Handle malloc(0) and malloc(__CPROVER_max_malloc_size) in assigns clause instrumentation

### PR DESCRIPTION
**Pull request by remi-delmas-3000**
With the previous upper bound expression`ub = lb + size - 1` pointer overflow were possible when the object size was zero.

The expression has been changed to `ub = lb + size` to allow for a zero size without overflows.

The CAR inclusion tests do not need to be changed.

Added tests to demonstrate that both `malloc(0)` and `malloc(__CPROVER_max_malloc_size)` are handled without issue by CAR instrumentation and contract replacement.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: regression/contracts/assigns-enforce-malloc-zero/test.desc [hunk 1]
```diff
@@ -0,0 +1,11 @@
+CORE
+main.c
+--enforce-contract foo _ --malloc-may-fail
+^EXIT=0$
+^SIGNAL=0$
+^\[foo\.\d+\] line 22 Check that a\[\(signed long int\)i\] is assignable: SUCCESS
```
Comment (tautschnig) [inline regression/contracts/assigns-enforce-malloc-zero/test.desc]: Use `signed (long (long )?)?int` to ensure this works across several platforms.

File: regression/contracts/assigns-enforce-malloc-zero/test.desc [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+^VERIFICATION SUCCESSFUL$
+--
+This test checks that object of size zero or of __CPROVER_max_malloc_size
```
Comment (tautschnig) [inline regression/contracts/assigns-enforce-malloc-zero/test.desc]: Nit: "objects" (plural)

File: regression/contracts/assigns-enforce-malloc-zero/main.c [hunk 3]
```diff
@@ -0,0 +1,34 @@
+#include <stdlib.h>
+
+int MAX_SIZE = __CPROVER_max_malloc_size;
```
Comment (tautschnig) [inline regression/contracts/assigns-enforce-malloc-zero/main.c]: Nit pick: I'm not even sure why this is necessary, but if it's just about having a shorter name then a `#define` would seem more idiomatic?

File: regression/contracts/assigns-replace-malloc-zero/main.c [hunk 4]
```diff
@@ -0,0 +1,40 @@
+#include <stdlib.h>
+
+int MAX_SIZE = __CPROVER_max_malloc_size;
```
Comment (tautschnig) [inline regression/contracts/assigns-replace-malloc-zero/main.c]: See comment above.

File: src/goto-instrument/contracts/assigns.cpp [hunk 5]
```diff
@@ -133,8 +133,7 @@ assigns_clauset::conditional_address_ranget::generate_snapshot_instructions()
 
   instructions.add(goto_programt::make_assignment(
     upper_bound_address_var,
-    minus_exprt{plus_exprt{slice.first, slice.second},
-                from_integer(1, slice.second.type())},
+    plus_exprt{slice.first, slice.second},
```
Comment (tautschnig) [inline src/goto-instrument/contracts/assigns.cpp:136]: Why was the "-1" necessary before, or why is this the only change that's necessary?

**Review by tautschnig — COMMENTED** [review]
I'm not yet clear why there wasn't any need to change a comparison operator in addition to the single arithmetic change. Could you please clarify?

File: regression/contracts/assigns-enforce-malloc-zero/main.c (hunk 3 above)
Comment (jimgrundy) [reply regression/contracts/assigns-enforce-malloc-zero/main.c]: I'm allergic to macros, but if its just about having a shorter name then I'd go with "const int MAX_SIZE = ...;"

File: src/goto-instrument/contracts/assigns.cpp (hunk 5 above)
Comment (jimgrundy) [reply src/goto-instrument/contracts/assigns.cpp:136]: I guess we were expecting some "x <= y" checks to become "x < y" checks, so I too am curious to know why that didn't need to happen.

Comment (TGWDB) [reply src/goto-instrument/contracts/assigns.cpp:136]: I'd also be curious to know why changing the actual value used does not appear to break anything. Was there a bug before, or is there an edge case now not being tested?

File: regression/contracts/assigns-enforce-malloc-zero/test.desc (hunk 1 above)
Comment (remi-delmas-3000) [reply regression/contracts/assigns-enforce-malloc-zero/test.desc]: fixed

**Comment by remi-delmas-3000** [comment]
> I'm not yet clear why there wasn't any need to change a comparison operator in addition to the single arithmetic change. Could you please clarify?

these expressions define upper bounds for intervals. We just interpret them as being included/excluded of the interval, but the inclusion check for two intervals just compares the upper bounds using `<=`

File: src/goto-instrument/contracts/assigns.cpp (hunk 5 above)
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/assigns.cpp:136]: these expressions define upper bounds for intervals. We just interpret them as being included/excluded of the interval, but the inclusion check for two intervals just compares the upper bounds using `_ <= _`. Our test suite was missing a case with objects of size 0 and `__CPROVER_max_malloc_size`. My only fear removing the `-1` would be a pointer overflow into the object bits, but it seems like the fact that pointers pointing to one past an object's boundary are legal in C is already factored in `__CPROVER_max_malloc_size`, so everything works.

File: regression/contracts/assigns-enforce-malloc-zero/main.c [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+int nondet_int();
```
Comment (feliperodri) [inline regression/contracts/assigns-enforce-malloc-zero/main.c]: Why do you need this `nondet_int()`?

File: regression/contracts/assigns-enforce-malloc-zero/main.c [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  __CPROVER_requires(0 <= size && size <= MAX_SIZE)
+  __CPROVER_requires(a == NULL || __CPROVER_is_fresh(a, size))
+  __CPROVER_assigns(__CPROVER_POINTER_OBJECT(a))
+  __CPROVER_ensures(
+    a && __CPROVER_return_value >= 0 ==> a[__CPROVER_return_value] == 0)
+// clang-format on
+{
+  if(!a)
+    return -1;
+  int i = nondet_int();
```
Comment (feliperodri) [inline regression/contracts/assigns-enforce-malloc-zero/main.c]: If you leave it uninitialized it'll have a nondeterministic value.

File: regression/contracts/assigns-enforce-malloc-zero/test.desc [hunk 8]
```diff
@@ -0,0 +1,11 @@
+CORE
+main.c
+--enforce-contract foo _ --malloc-may-fail
```
Comment (feliperodri) [inline regression/contracts/assigns-enforce-malloc-zero/test.desc]: You should also include `--malloc-fail-null`.

File: regression/contracts/assigns-replace-malloc-zero/main.c [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+int nondet_int();
```
Comment (feliperodri) [inline regression/contracts/assigns-replace-malloc-zero/main.c]: Same as above.

File: regression/contracts/assigns-replace-malloc-zero/main.c [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+__CPROVER_requires(0 <= size && size <= MAX_SIZE)
+__CPROVER_requires(a == NULL || __CPROVER_rw_ok(a, size))
+__CPROVER_assigns(__CPROVER_POINTER_OBJECT(a))
+__CPROVER_ensures(
+    a && __CPROVER_return_value >= 0 ==> a[__CPROVER_return_value] == 0)
+// clang-format on
+{
+  if(!a)
+    return -1;
+  int i = nondet_int();
```
Comment (feliperodri) [inline regression/contracts/assigns-replace-malloc-zero/main.c]: Same as above.

File: regression/contracts/assigns-replace-malloc-zero/test.desc [hunk 11]
```diff
@@ -0,0 +1,11 @@
+CORE
+main.c
+--replace-call-with-contract foo _ --malloc-may-fail
```
Comment (feliperodri) [inline regression/contracts/assigns-replace-malloc-zero/test.desc]: Same as above.

File: regression/contracts/assigns-replace-malloc-zero/test.desc [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+^EXIT=0$
+^SIGNAL=0$
+\[main\.assertion\.1\] line 38 expecting SUCCESS: SUCCESS$
+^VERIFICATION SUCCESSFUL$
+--
+This test checks that object of size zero or of __CPROVER_max_malloc_size
```
Comment (feliperodri) [inline regression/contracts/assigns-replace-malloc-zero/test.desc]: objects

**Review by feliperodri — CHANGES_REQUESTED** [review]

File: regression/contracts/assigns-enforce-malloc-zero/test.desc (hunk 2 above)
Comment (remi-delmas-3000) [reply regression/contracts/assigns-enforce-malloc-zero/test.desc]: fixed

File: regression/contracts/assigns-enforce-malloc-zero/main.c (hunk 3 above)
Comment (remi-delmas-3000) [reply regression/contracts/assigns-enforce-malloc-zero/main.c]: fixed

File: regression/contracts/assigns-replace-malloc-zero/main.c (hunk 4 above)
Comment (remi-delmas-3000) [reply regression/contracts/assigns-replace-malloc-zero/main.c]: fixed

File: regression/contracts/assigns-enforce-malloc-zero/test.desc (hunk 8 above)
Comment (remi-delmas-3000) [reply regression/contracts/assigns-enforce-malloc-zero/test.desc]:
```
 --malloc-may-fail          allow malloc calls to return a null pointer
 --malloc-fail-null           set malloc failure mode to return null
 ```
 `--malloc-may-fail` seems more general than `--malloc-fail-null`, isn't it ?

File: regression/contracts/assigns-replace-malloc-zero/main.c (hunk 9 above)
Comment (remi-delmas-3000) [reply regression/contracts/assigns-replace-malloc-zero/main.c]: simplified

File: regression/contracts/assigns-enforce-malloc-zero/main.c (hunk 7 above)
Comment (remi-delmas-3000) [reply regression/contracts/assigns-enforce-malloc-zero/main.c]: simplified

File: regression/contracts/assigns-enforce-malloc-zero/test.desc (hunk 8 above)
Comment (feliperodri) [reply regression/contracts/assigns-enforce-malloc-zero/test.desc]: I'd expect the same thing, but you need the second flag to ensure that `malloc` will return NULL on failure. See the following example.
```c
// file main.c
#include <stdlib.h>

int main()
{
  int *i = malloc(0);
  assert(__CPROVER_DYNAMIC_OBJECT(i));
  //assert(i == NULL);
}
```

If you run it using `cbmc main.c --malloc-may-fail` the result is:
```shell
** Results:
<builtin-library-malloc> function malloc
[malloc.assertion.1] line 26 max allocation size exceeded: SUCCESS
[malloc.assertion.2] line 31 max allocation may fail: SUCCESS

main.c function main
[main.assertion.1] line 6 assertion IS_DYNAMIC_OBJECT((const void *)i): SUCCESS

** 0 of 3 failed (1 iterations)
VERIFICATION SUCCESSFUL
```

But if you run using `cbmc main.c --malloc-may-fail  --malloc-fail-null --trace`, the result is:
```shell
** Results:
<builtin-library-malloc> function malloc
[malloc.assertion.1] line 26 max allocation size exceeded: SUCCESS
[malloc.assertion.2] line 31 max allocation may fail: SUCCESS

main.c function main
[main.assertion.1] line 6 assertion IS_DYNAMIC_OBJECT((const void *)i): FAILURE

Trace for main.assertion.1:

State 26 file main.c function main line 5 thread 0
----------------------------------------------------
  i=((signed int *)NULL) (00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000)

State 30 file main.c function main line 5 thread 0
----------------------------------------------------
  malloc_size=0ul (00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000)

State 41 file main.c function main line 5 thread 0
----------------------------------------------------
  i=((signed int *)NULL) (00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000)

Violated property:
  file main.c function main line 6 thread 0
  assertion IS_DYNAMIC_OBJECT((const void *)i)
  IS_DYNAMIC_OBJECT((const void *)i)



** 1 of 3 failed (2 iterations)
VERIFICATION FAILED
```

Comment (remi-delmas-3000) [reply regression/contracts/assigns-enforce-malloc-zero/test.desc]: Fixed. 

Remark: The CLI help and doc should be updated to reflect the actual behaviour.

File: regression/contracts/assigns-replace-malloc-zero/test.desc (hunk 12 above)
Comment (remi-delmas-3000) [reply regression/contracts/assigns-replace-malloc-zero/test.desc]: fixed

**Review by feliperodri — APPROVED** [review]
